sui_core/src/safe_client.rs

```diff
@@ -310,39 +313,61 @@ where
             .handle_batch_stream(request.clone())
             .await?;
 
-        let seq_requested = request.end - request.start;
-        let mut seq_to_be_returned = seq_requested as usize;
-        // check for overflow
-        if seq_requested > usize::MAX as u64 {
-            seq_to_be_returned = usize::MAX;
-        }
-
         let client = self.clone();
-        let stream = Box::pin(
-            batch_info_items
-                .then(move |batch_info_item| {
-                    let req_clone = request.clone();
-                    let client = client.clone();
-                    async move {
-                        match &batch_info_item {
-                            Ok(BatchInfoResponseItem(UpdateItem::Batch(signed_batch))) => {
-                                if let Err(err) =
-                                    client.check_update_item_batch_response(req_clone, signed_batch)
-                                {
-                                    client.report_client_error(err.clone());
-                                    return Err(err);
-                                }
-                                batch_info_item
-                            }
-                            Ok(BatchInfoResponseItem(UpdateItem::Transaction((_seq, _digest)))) => {
-                                batch_info_item
-                            }
-                            Err(e) => Err(e.clone()),
+        let address = self.address;
+        let stream = Box::pin(batch_info_items.scan(
+            (0u64, None),
+            move |(seq, txs_and_last_batch), batch_info_item| {
+                let req_clone = request.clone();
+                let client = client.clone();
+
+                // We check if we have exceeded the batch boundary for this request.
+                if !(*seq < request.end) {
+                    // If we exceed it return None to end stream
+                    return futures::future::ready(None);
+                }
+
+                let x = match &batch_info_item {
+                    Ok(BatchInfoResponseItem(UpdateItem::Batch(signed_batch))) => {
+                        println!("batch {:?}", signed_batch.batch);
+                        if let Err(err) = client.check_update_item_batch_response(
+                            req_clone,
+                            &signed_batch,
+                            &txs_and_last_batch,
```

Comment:
Does this need to be populated with the transactions from the batch first? I think this is what confused me before, that we don't store the digests of each transaction in the batch. If we use all the digests to create the batch, but then we don't pass in all the digests to reconstruct the batch, is it still possible we get the same batch after reconstruction?